crypto/x509/v3_ncons.c

```diff
@@ -782,50 +783,56 @@ static int nc_email(ASN1_IA5STRING *eml, ASN1_IA5STRING *base)
 static int nc_uri(ASN1_IA5STRING *uri, ASN1_IA5STRING *base)
 {
     const char *baseptr = (char *)base->data;
-    const char *hostptr = (char *)uri->data;
-    const char *p = ia5memchr(uri, (char *)uri->data, ':');
+    char *uri_copy;
+    char *scheme;
+    char *host;
     int hostlen;
+    int ret;
 
-    /* Check for foo:// and skip past it */
-    if (p == NULL
-            || IA5_OFFSET_LEN(uri, p) < 3
-            || p[1] != '/'
-            || p[2] != '/')
-        return X509_V_ERR_UNSUPPORTED_NAME_SYNTAX;
-    hostptr = p + 3;
-
-    /* Determine length of hostname part of URI */
+    if ((uri_copy = OPENSSL_strndup((const char *) uri->data, uri->length)) == NULL)
+        return X509_V_ERR_UNSPECIFIED;
 
-    /* Look for a port indicator as end of hostname first */
+    if (!OSSL_parse_url(uri_copy, &scheme, NULL, &host, NULL, NULL, NULL, NULL, NULL)) {
+        OPENSSL_free(uri_copy);
+        return X509_V_ERR_UNSUPPORTED_NAME_SYNTAX;
+    }
 
-    p = ia5memchr(uri, hostptr, ':');
-    /* Otherwise look for trailing slash */
-    if (p == NULL)
-        p = ia5memchr(uri, hostptr, '/');
+    /* Make sure the scheme is there */
+    if (scheme == NULL) {
+        fprintf(stderr, "x509: missing scheme in URI: %s\n", uri_copy);
```

Comment:
(BTW, I agree that OpenSSL diagnostics should be extended in many cases.
And adding nice diagnostics would better involve `ERR_raise_data(...)` than `fprintf(stderr, ..)` 
but turns out not to be needed here.)

Reply:
Ah, I was wondering if there was a better way than just writing to `stderr`... thanks!